src/dmd/statement.d

```diff
@@ -449,6 +449,88 @@ extern (C++) abstract class Statement : ASTNode
     inout(CompoundDeclarationStatement) isCompoundDeclarationStatement() { return stmt == STMT.CompoundDeclaration ? cast(typeof(return))this : null; }
 }
 
+
+private Statement stripScopeStatement(Statement s)
+{
+    Statement result = s;
+
+    if (s)
+    {
+        ScopeStatement ss = s.isScopeStatement();
+        while(ss && ss.statement)
+        {
+            result = ss.statement;
+            ss = result.isScopeStatement();
+        }
+    }
+
+    return result;
+}
+/**
+* Returns:
+*     Wheter ($D stmt) containts assert(ctfe)
+* Params:
+*  stmt = CompoundStatement to check
+*/
+bool containsAssertCtfe(CompoundStatement stmt)
+{
+    bool result = false;
+
+    Statements *statements;
+    if (stmt !is null)
+    {
+        if (CompoundStatement cs = stmt.isCompoundStatement())
+        {
+            statements = cs.statements;
+        }
+    }
+
+    if (statements !is null)
+    {
+        foreach(s;(*statements))
+        {
+            if (!s)
+                continue;
+
+            s = stripScopeStatement(s);
+            if (isAssertCtfe(s))
```

Comment:
One question about the design here (it touches on my other remark about test cases).

The current functionality is based on finding an `assert(__ctfe)` statement, right?  Whether in the function body or an in-contract?

However ... will the correct behaviour also result if someone uses the newer in-contract syntax of just `in(__ctfe)` ... ?  If not, is it worth adding support for that too?

Finally, can the same result be obtained _without_ any compiler feature change, by asking the user to use a conditional, e.g.

```D
string foo (string x, string y)
    if (__ctfe)
{
    return x ~ " " ~ y;
}
```

... ?

Reply:
It will automatically support the new in syntax.
As that is rewritten in the parser.

The `if (__ctfe)` is impossible to do I am afraid.

Reply:
>  As that is rewritten in the parser.

Cool.  Just needs a test case then :-)

> The `if (__ctfe)` is impossible to do I am afraid.

I wasn't sure.  Thanks for clarifying :-)